### ISSUE TITLE
hs.window.filter and AX roles for windows spawned by HS

### ISSUE THREAD

**Issue by lowne** [Hammerspoon/hammerspoon#576]
The premise: `hs.window.filter` should completely disregard container windows for `hs.drawing` and `hs.alert` objects; since at any given moment there could be tons of drawings (and, to a lesser extent, alerts) being created/destroyed (e.g. `hs.grid.show()`), this is for performance reasons, and to avoid cluttering the log.

This used to be the case until 0.9.40 - drawing windows, just like alerts, had a `:subrole()` of `AXUnknown` so it was easy to filter them out.

With recent changes (I suppose having to do with activating vs not activating HS) it seems that drawings now are `AXSystemDialog`, at least for rects (but I haven't checked all different the ways to create drawings). To complicate things further, `hs.webview` windows (which we do _not_ want to filter) can have varying subroles; what I have found so far is `AXUnknown` for `windowStyle(0)`, `AXDialog` when titled, and `AXFloatingWindow` for "utility".

So:
1. I assume these AX subroles are somehow determined by the OS, but just to be sure, is there no way to assign them directly?
2. If not, is there a way to have drawings behave as desired, but still being `AXUnknown`? Do alerts (which still are AXUnknown) exhibit the same problem? If not, can whatever method is used for them be also used for drawings?
3. If not, is it possible to find a stable way so that drawing containers are _always_, say, `AXSystemDialog`, and webviews are _never_ that?
4. Any other way to disambiguate drawings/alerts/webviews/HS Console/HS Preferences?

(To be clear, if we can get back to AXUnknown for everything, I'm ok with "sacrificing" title-less AXUnknown webviews since they should be "background", or at best transient, windows - in that case I'll just mention in the docs for windowfilter to use titled webviews if the user wants to be able to address them)

(ping @asmagill)

**Comment by asmagill** [comment]
The AX roles and subroles are, as far as I can determine, set by the system and not directly changeable by us.  There are some patterns as shown in pull #566, though I can try and get a more formal list this weekend if you like.  I'll also dig a little deeper into NSPanel and see if we can identify them another way.

**Comment by lowne** [comment]
Ah, I had missed your reply in #566 for some reason, maybe because it's a PR. Looking at that I think alternative 3 is hopeless, and since drawings also have all sorts of windowBehaviors and windowLevels the distinction apparently cannot be between drawings and webviews tout court, but between "proper or semi-proper windows" (which could include clickCallback drawings, title-bar webviews, etc.) and everything else, which seems to point right back at just AXUnknown.

So: wrt PR #567 (which changed the drawings): is it possible to have it so that *by default* hs.drawing objects end up being AXUnknown, as they used to?

As an example, imagine a tinted overlay over the screen (`hs.grid.show()` is a good example of the basic usage for non-interactive drawings); with just `hs.drawing.rect():setFillColor()` *and nothing else*, the resulting drawing should
- have *behaviour* `canJoinAllSpaces` or `moveToActiveSpace` (not sure which one is better, but surely not `default`), and `transient`
- have *windowLevel* of `floating` or `overlay` (probably? not sure what they all do)
- result in a subrole of `AXUnknown`

If and only if the user changes these or calls `:setClickCallback()` should the subrole change - and it's acceptable (and maybe appropriate) to have those not filtered.

Since #567 was about clickable drawings, it should be possible to arrange things such that non-clickable ones aren't affected; in the absolute worst case, calling :setClickCallback would have to tear down the previous window (NSWindow) and create a new one (NSPanel), but I'm sure that won't be necessary, because, as shown by webviews, NSPanels can be AXUnknown as well - just delay whatever change is required for the non-activating behaviour until :setClickCallback is actually called.

**Comment by lowne** [comment]
(another reason to prefer going back to AXUnknown, or more specifically to avoid AXSystemDialog for interactionless overlays, is that AXSystemDialog drawings break my "focusModalDialog" windowfilter, which I can't live without anymore - between HS, Vimari for Safari, and a lot of shortcuts, I'm practically keyboard-only now)

**Comment by asmagill** [comment]
Ok, let's take this one step at a time... looking at hs.drawing and ignoring hs.webview for the moment...  Webview has more possible options (titled, resizable, panel, hud, etc.) that are not available as options in drawing, so let's get drawing sorted first.


Since you mentioned it an I don't know, does the setting of windowBehavior change the role and subrole values?  Limiting ourselves to the style only (which for drawing means non-activating or not, since those are the only styles currently available in drawing) we have the following:

If using NSWindow, we get (activating is the only possible option):
    0 0000000000000000    AXWindow             AXUnknown

as the default (only)


If using NSPanel, the current default would be (this is non-activating):
  128 0000000010000000    AXWindow             AXSystemDialog

I can easily add a toggle so that the following is also available or even the default (this is activating):
    0 0000000000000000    AXWindow             AXUnknown

Again, this assumes that behavior doesn't affect things further, so let me know if your experience suggests it does.

If these are the only patterns we need to worry about, would it be sufficient to default to the second panel option and allow a toggle to change it?  If your console is usually closed or you don't mind HS coming forward, there is no need for clickCallback to force it... make the user choose it if (like me) they specifically want it.

--
A-Ron

**Comment by asmagill** [comment]
I have an updated hs.drawing with a method :clickCallbackActivating, default true, based upon my previous post... if that matches your requirements for hs.drawing, I can create the pull request and get it installed.  Let me know.

**Comment by lowne** [comment]
That should work, but as I said it would probably be appropriate to allow clickable drawings by default in the root filter; so
- new drawings are created activating (i.e. AXUnknown), because they don't activate anything anyway and we want them filtered
- as soon as you call :setClickCallback on them they would become non-activating (which is the sensible thing) and therefore AXSystemDialog (which we want as well, so they're not filtered)

But yes, we should probably have a way of addressing those rare (if any at all) cases where the user or a module want things the other way around - which would be for the sole reason of keeping the AXUnknown subrole for windowfilter-like uses; since this would only make sense for clickable drawings, and be a no-op otherwise, it'd probably be better to have it as a 3rd argument (say, `noSystemDialog`) to setClickCallback instead of a separate clickCallbackActivating method.

To recap:
- `mydrawing=hs.drawing.a_constructor(frame,...)` creates an activating NSPanel, AXUnknown
- `mydrawing:setClickCallback(upfn,downfn)` makes the NSPanel non-activating, AXSystemDialog
- but `mydrawing:setClickCallback(upfn,downfn,true)` doesn't, so the NSPanel is still activating and AXUnknown

**Comment by asmagill** [comment]
I've not really looked that closely at hs.window.filter so at present, I'm not using it -- do you want the filter to catch clickable drawings?  We're talking about drawings here, not webview windows; clickable drawings have no additional title or close button or anything -- the only difference between them and a grid line being drawn in hs.grid is that mouse clicks don't pass through to the window underneath... if I'm reading you correctly in your last post, you *do* want the filter to match these, not skip them by default, correct?

**Comment by lowne** [comment]
Correct; the idea was that users *might* conceivably want to do something with interactive (aka clickable) drawings' "windows" (yes, titleless and all), but not for "cosmetic" drawings; however, thinking about it some more, I changed my mind - I think we should let the user decide in both cases; as a contrived example I thought of having a set of always-on-top "status" drawings close to the top right corner: whether they're actionable (clickable) or not, it's possible that the user might want to use window.layout (which relies on window.filter) on them to have them correctly placed as external screens are attached/detached. So, disregard my previous message.

A generalised solution might be:

1. add `hs.drawing:setStyle()`/`:style()` methods, mirroring `:setBehavior()` (or maybe `:windowStyle()` like webviews - I suppose it'd be nice to have consistency) - which would also allow drawings to be titled, closable, resizable etc. (could be an idea for mouse-enabled hs.expose)
2. have the docstring for `:setClickCallback()` mention the possibility of further calling `:setStyle(styles.nonActivating)` to avoid the console-activating thing
3. have the docstring for `:setStyle()` note that setting `nonActivating`, `titled` etc changes the subrole from `AXUnknown` to `AXSystemDialog` (or whatever else might be the case)
4. likewise, I'll change the docstrings in windowfilter to mirror this - something along the lines of "windows created by Hammerspoon - alerts, drawings and webviews - by default have a role of 'AXUnknown' and are filtered at the root. If you want to be able to address webviews or drawings, use `:setStyle(...)` on them".

This way we'd have maximum flexibility for both user configs and HS modules using drawings.

**Comment by asmagill** [comment]
@lowne, was the original issue corrected with the last changes to hs.drawing?  Can we close this or are there still issues to work out?

**Comment by asmagill** [comment]
@lowne, I have recently found that setting the accessibility subrole to a custom value is actually extremely simple... Is this still something that would be useful for identifying `hs.drawing` objects?

**Comment by lowne** [comment]
Yes, it would make it trivial to exclude drawings from window filters

**Comment by asmagill** [comment]
I've made the change in master for hs.drawing... initial experiments with hs.webview and some potential new modules I'm considering for a future release suggest that certain settings to NSWindow objects will reset subrole back to one of the standard values, but since the window portion of drawings doesn't change, this update should work for your current purposes.

Hammerspoon drawing objects will have the subrole of "hammerspoonDrawing" in the next release.

**Comment by cmsj** [comment]
I'm going to close this because I'm gardening the bugs and this has been open for ages with no further discussion, feel free to re-open it if there are specific things you want to do :)
